File: design_patterns/002_observer/group_python/WeatherStation/src/DisplayStatistic.py
from IObserver import IObserver
from IDisplay import IDisplay
from IObservable import IObservable
from WeatherObservable import WeatherInfo


class DisplayStatistic(IObserver, IDisplay):
    _weather_data:IObservable
    _max_temp:float = 0.0
    _min_temp:float = 200.0
    _temp_sum:float = 0.0
    _num_readings:int = 0


    def __init__(self, weather_data:IObservable) -> None:
        self._weather_data = weather_data
        self._weather_data.substribe(self)

    def update(self, updated_data: dict) -> None:
        self._num_readings += 1
        self._temp_sum += updated_data.get(WeatherInfo.TEMPERATURE)
        
        self._max_temp = max(updated_data.get(WeatherInfo.TEMPERATURE), self._max_temp)
        self._min_temp = min(updated_data.get(WeatherInfo.TEMPERATURE), self._min_temp)
        self.display() 

    def display(self) -> None:
        print(f'Avg/Max/Min temperature = {self._temp_sum/self._num_readings}/{self._max_temp}/{self._min_temp}')


if __name__ == '__main__':
    raise Exception(f'This module should not be executed directly. Only for imports')

File: design_patterns/002_observer/solutions/typescript/weatherStationRxJs2/src/DisplayStatistic.ts
import { Observable } from "rxjs";
import { IDisplay } from "./IDisplay";
import { WeatherInfoType } from "./WeatherData";

export class DisplayStatistic implements IDisplay {
  private maxTemp: number = 0.0;
  private minTemp: number = 200.0;
  private tempSum: number = 0.0;
  private numReadings: number = 0;

  constructor(private weatherData: Observable<WeatherInfoType>) {
    this.weatherData.subscribe(data => this.update(data));
  }

  update(updatedData: WeatherInfoType): void {
    const currentTemp = updatedData.temperature;

    if (currentTemp !== undefined) {
      this.tempSum += currentTemp;
      this.numReadings++;

      if (currentTemp > this.maxTemp) {
        this.maxTemp = currentTemp;
      }

      if (currentTemp < this.minTemp) {
        this.minTemp = currentTemp;
      }

      this.display();
    }
  }

  display(): void {
    const avgTemp = this.tempSum / this.numReadings;
    console.log(`Avg/Max/Min temperature = ${avgTemp}/${this.maxTemp}/${this.minTemp}`);
  }
}
